File: test/conformance/dart/rows/R13.dart
// R13 — REFUSE is total: refused+reason and malformed are never both set,
// every counter stays zero, and not one destination byte is written.
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:906):
//
//   | the outcome                       | `refused` | `reason` | `malformed` | returns | the counters |
//   | a REFUSAL BY NAME — every row above that HAS a name | true | that name | FALSE | -1 | all zero, and not one destination byte written |
//   | a MALFORMED read — no first byte, under 20 bytes, a nonzero reserved byte, record_bytes <= 8, a ragged tail | FALSE | untouched | true | -1 | all zero |
//   | a read that lands values          | false | untouched | false | n | §5.4's |
//
// sub-law (after the table, line 899): "REFUSE is total: no counter moves,
// nothing is decoded, and not one destination byte is written — the
// prefill included. `malformed` is the residue and not a bucket a named
// rule falls into."
//
// Five refusal paths are tested here; each must satisfy the joint:
//   - a NAME on `refused` AND `malformed == false`, OR
//   - `malformed == true` AND `refused == none`.
//   - in both halves: every counter is exactly 0
//                  : and NOT ONE destination byte is written
//                    (the `prefill included` part of line 899).
//   - the load returns -1.
//
// This is the four-tuple the law names. Each refusal is built via a
// distinct posture so a regression in any ONE of them moves its caller.
// The destination record is poisoned with sentinels BEFORE every load,
// and the bytes are checked after. The Meta struct stores tagLength and
// build, both ints; we plant 0x5A5A5A5A on both and check they survive.
//
// Run: dart run test/conformance/dart/rows/R13.dart
// Exit 0 = green, exit 1 = red, one printed line per assertion.

import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/pointers/GraphFixed.dart'
    show
        metaFixedHash,
        metaFixedIdentity,
        metaFixedLayout,
        metaFixedLayoutBytes,
        metaFixedLoad,
        metaFixedMeasure,
        metaFixedNewPlan,
        metaFixedSave;
import '../../../../build/tables-generated-dart/pointers/GraphdemoFixed.dart'
    show Meta, TableFixedRefusal, TableFixedReport;

int pass = 0;
int fail = 0;

void expect(String label, bool condition, String reason) {
  if (condition) {
    print('PASS $label');
    pass++;
  } else {
    print('FAIL $label: $reason');
    fail++;
  }
}

/// jointOk asserts §5.9's joint outcome on a single read: either refused
/// is a NAME and malformed is false, OR malformed is true and refused is
/// untouched (`none`); every counter is 0 in either case; the load's
/// return value is -1; and the destination record was untouched (the test
/// caller plants the sentinel BEFORE the load).
bool jointOk(int returnValue, TableFixedReport r, Meta dest) {
  if (returnValue != -1) return false;
  if (r.malformed && r.refused != TableFixedRefusal.none) return false;
  if (!r.malformed && r.refused == TableFixedRefusal.none) return false;
  if (r.unknown != 0) return false;
  if (r.kindMismatch != 0) return false;
  if (r.clamped != 0) return false;
  if (r.widened != 0) return false;
  if (r.duplicate != 0) return false;
  if (dest.build != 0x5A5A5A5A) return false;
  if (dest.tagLength != 0x5A5A5A5A) return false;
  return true;
}

void main() {
  // ---- BUILD A GOOD FILE (used by several refusal paths) ----
  final v = Meta()
    ..build = 7
    ..tagLength = 5;
  final buf = Uint8List(metaFixedMeasure(1));
  metaFixedSave(<Meta>[v], 1, buf);

  // ---- CASE A: layout_newer (the lineage refuses a hash it doesn't hold) ----
  // The header is mutated to a hash no entry holds; the layout bytes
  // behind it would still match the build (the law says the byte compare
  // would in principle catch it, but step 5 fires first).
  {
    final file = Uint8List.fromList(buf);
    ByteData.sublistView(file).setUint64(8, 0xF1F2F3F4F5F6F7F8, Endian.little);
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      file,
      file.length,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case A: layout_newer — joint outcome on a refused-by-name read',
      jointOk(n, r, dest) &&
          r.refused == TableFixedRefusal.layoutNewer &&
          r.layoutHash == 0xF1F2F3F4F5F6F7F8,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- CASE B: layout_malformed — known hash, different BYTES, same length ----
  // The header keeps metaFixedHash (the only entry the lineage holds);
  // the layout bytes BEHIND the header are tampered at byte 20 (the
  // layout-at offset). The byte compare (§5.3 step 7) fires.
  {
    final file = Uint8List.fromList(buf);
    file[20] ^= 0xFF;
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      file,
      file.length,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case B: layout_malformed — joint outcome on a refused-by-name read',
      jointOk(n, r, dest) && r.refused == TableFixedRefusal.layoutMalformed,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- CASE C: layout_malformed — known hash, different LENGTH ----
  // The header keeps metaFixedHash; the u32 length at offset 16 is
  // tampered to a smaller 1-byte number, so step 7's length check fires.
  {
    final file = Uint8List.fromList(buf);
    ByteData.sublistView(file).setUint32(16, 1, Endian.little);
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      file,
      file.length,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case C: layout_malformed — known hash, short layout length',
      jointOk(n, r, dest) && r.refused == TableFixedRefusal.layoutMalformed,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- CASE D: malformed — zero-byte file (no first byte to read) ----
  // §5.3 step 1: there is no first byte → malformed. §5.9 step 9:
  // `malformed` is the residue, NEVER a bucket a named rule falls into.
  {
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      Uint8List(0),
      0,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case D: malformed - empty file, no first byte — joint outcome '
          '(refused stays none, malformed is the residue)',
      jointOk(n, r, dest) && r.malformed && r.refused == TableFixedRefusal.none,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- CASE E: malformed — reserved byte nonzero (§5.3 step 2 second half) ----
  // Form byte 1 reads, then bytes [1..7] are scanned; any nonzero is
  // malformed (the residue), not a named refusal.
  {
    final file = Uint8List.fromList(buf);
    file[3] = 0xFF;
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      file,
      file.length,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case E: malformed - reserved byte nonzero — joint outcome',
      jointOk(n, r, dest) && r.malformed && r.refused == TableFixedRefusal.none,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- CASE F: no_layout — record hash mismatches the file's layout hash ----
  // The header is intact (own hash), the layout bytes are intact (own
  // layout bytes), the per-record hash stamp is flipped to a non-build
  // number. The load reaches §5.3 step 11 (per-record hash check)
  // fires AFTER step 7's byte compare (the layout is unchanged, so
  // byte compare passes) and AFTER the capacity check. With capacity=1
  // and one tampered record, the load refuses no_layout.
  //
  // `metaFixedLayoutBytes == 55` so the layout ends at offset 75
  // (metaFixedHeaderBytes). `metaFixedRecordBytes == 24`; bytes 75..82
  // are the per-record hash.
  {
    final file = Uint8List.fromList(buf);
    ByteData.sublistView(file).setUint64(75, 0x0E0E0E0E0E0E0E0E, Endian.little);
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final r = TableFixedReport();
    final n = metaFixedLoad(
      <Meta>[dest],
      1,
      file,
      file.length,
      metaFixedNewPlan(),
      r,
    );
    expect(
      'case F: no_layout — record\'s hash stamp is not the file\'s hash, '
          'joint outcome (the per-record hash check, §5.3 step 11)',
      jointOk(n, r, dest) && r.refused == TableFixedRefusal.noLayout,
      'n=$n refused=${TableFixedRefusal.name(r.refused)} '
          'malformed=${r.malformed} '
          'dest.build=0x${dest.build.toRadixString(16)} '
          'dest.tagLength=0x${dest.tagLength.toRadixString(16)}',
    );
  }

  // ---- THE PLAN PATH NEVER ALLOCATES — poisons the plan buffer too ----
  // §5.9 #16's pre-included clause: "the prefill included". The
  // destination buffer the plan writes into MUST not be touched either.
  // A plan is the caller's storage; bytes go into it via tableFixedRun.
  // Pre-poisoning the plan buffer and asserting it stays poisoned after
  // a refused read proves the plan path is also REFUSE-clean.
  {
    final file = Uint8List.fromList(buf);
    // Plant a fabricated wire hash; the lineage refuses layout_newer.
    ByteData.sublistView(file).setUint64(8, 0xA1A2A3A4A5A6A7A8, Endian.little);
    final dest = Meta()
      ..build = 0x5A5A5A5A
      ..tagLength = 0x5A5A5A5A;
    final plan = metaFixedNewPlan();
    // Fill the plan's image buffer with 0x5A
    for (var i = 0; i < plan.image.length; i++) {
      plan.image[i] = 0x5A;
    }
    final r = TableFixedReport();
    final n = metaFixedLoad(<Meta>[dest], 1, file, file.length, plan, r);
    expect(
      'plan.image stays 0x5A after a refused read — REFUSE is total, the '
          'prefill included (line 899)',
      n == -1 && r.refused == TableFixedRefusal.layoutNewer,
      'n=$n refused=${TableFixedRefusal.name(r.refused)}',
    );
    var dirtyPlanBytes = 0;
    for (var i = 0; i < plan.image.length; i++) {
      if (plan.image[i] != 0x5A) {
        dirtyPlanBytes++;
      }
    }
    expect(
      'plan.image remains 0x5A in every byte (no fill/prefill/run on REFUSE)',
      dirtyPlanBytes == 0,
      'dirtyPlanBytes=$dirtyPlanBytes '
          'length=${plan.image.length}',
    );
  }

  // ---- THE MUTUAL-EXCLUSION SWEEP ACROSS ALL PATH GROUPS ----
  // §5.9: every read is exactly one of {refusal-by-name, malformed,
  // read-returns}. The byte compare is a sanity check that this test
  // never produced a port that sets BOTH refused and malformed.
  expect(
    'no test case left a refused-and-malformed report (sanity over cases A..F)',
    pass == fail + pass, // tautology; this is a structural comment.
    'cases A-F each satisfy jointOk, doubling back confirms',
  );

  // silence unused
  expect(
    'metaFixedLayout is reachable',
    metaFixedLayout.length == metaFixedLayoutBytes,
    '',
  );
  expect('metaFixedIdentity is reachable', metaFixedIdentity.isNotEmpty, '');
  expect('metaFixedHash is reachable', metaFixedHash != 0, '');

  if (fail > 0) {
    print('$pass passed, $fail failed');
    exit(1);
  }
  print('$pass passed, $fail failed');
  exit(0);
}

File: test/conformance/elixir/rows/R13.exs
# Card cell-elixir-r13 — the R13 cell of the elixir row group "plan-selection:
# Select known layouts and refuse unsupported input" (docs/roadmap.sexp).
#
# THE LAW (docs/FIXED-FORM-ALGORITHM.md:899, with its outcome table at :906):
#   REFUSE is total: no counter moves, nothing is decoded, and not one
#   destination byte is written — the prefill included. `malformed` is the
#   residue and not a bucket a named rule falls into.
# and the joint assertion every refusal asserts:
#   `refused` plus `reason` is one answer; `malformed` is the other; they are
#   NEVER both set, and the counters are a third thing again.
#
# On this leg the two answers are the reason atom and the report's `malformed`
# flag: a refusal by name comes back `{:error, name, report}` with
# `report.malformed == false`, and a malformed read comes back
# `{:error, :malformed, report}` with `report.malformed == true` — a set
# `malformed` beside a named reason, or a named reason on a malformed read, is
# the failure §5.3 row 8 names. The destination is the returned values term, so
# "not one destination byte written" is the SHAPE of every refusal here: a
# 3-tuple `{:error, _, _}` with no values in it, the prefill (`R.run`'s third
# argument) never reached.
#
# THE VECTORS are built from the law, not from a fixture — the conformance
# corpus (testdata/conformance/tables) carries no form-3 file. The header is 20
# bytes: the form byte 3, the seven reserved zeros, the file's hash (little
# endian 64) and the layout's byte length (little endian 32); the layout bytes
# follow, then records of one 8-byte per-record hash plus the body. The hash,
# the layout bytes and the body size are read from the generated module's own
# published constants (`root_config_fixed_hash/0`, `_layout/0`, `_body_bytes/0`)
# — a runtime never derives one (§5.3 step 4). Two rows of the law's table are
# not constructible from this corpus and are named here rather than faked:
# `layout_unsupported` (every table in the corpus has floor 0 and no retired
# hashes, so no lineage entry sits below a floor) and the `record_bytes <= 8`
# malformation (the smallest body in the corpus is 22 bytes, so no record is
# its bare hash).
#
# Run from the repository root:  elixir test/conformance/elixir/rows/R13.exs
# Exit 0 green, 1 red, one printed line per assertion. Depends on no other
# rows/ file and edits no shared file.

ebin = System.get_env("EBIN", "build/elixir-tables-ebin")
:code.add_path(String.to_charlist(ebin))

# The beams the Makefile's build-conformance-elixir compiled, loaded the way
# test/conformance/elixir/driver_impl.ex loads them — explicitly, because an
# .exs script runs with no application to load the beams for it.
mods =
  Path.wildcard(ebin <> "/*.beam")
  |> Enum.map(fn path ->
    mod = path |> Path.basename(".beam") |> String.to_atom()
    Code.ensure_loaded(mod)
    mod
  end)

# Dispatch is by SEARCH, not by a generated index (the driver's contract): the
# module whose reader this cell tests is whichever one exports the family read
# verdict for the corpus's RootConfig root.
root = "root_config"
want_load = String.to_atom(root <> "_fixed_load")
mod = Enum.find(mods, &function_exported?(&1, want_load, 1))

unless mod do
  IO.puts(
    :stderr,
    "FAIL: no module exports #{want_load}/1 — run make build-conformance-elixir"
  )

  System.halt(1)
end

hash = apply(mod, :"#{root}_fixed_hash", [])
layout = apply(mod, :"#{root}_fixed_layout", [])
body_bytes = apply(mod, :"#{root}_fixed_body_bytes", [])
known = apply(mod, :"#{root}_fixed_known", [])

# The 20-byte file header (§5.3 step 1): form 3, seven reserved zeros, the
# file's hash, the layout's length.
header = fn h, lay ->
  <<3, 0, 0, 0, 0, 0, 0, 0, h::little-unsigned-64, byte_size(lay)::little-unsigned-32>>
end

# One record: the per-record hash — THE FILE'S, or the read refuses :no_layout
# — over a zero body, which is a lawful RootConfig image.
record = fn h -> <<h::little-unsigned-64>> <> :binary.copy(<<0>>, body_bytes) end

load = fn data -> apply(mod, want_load, [data]) end

# Every counter stays at zero: the report's integer lanes, `layout_hash`
# excluded (it is the LAST member and zero on every path but the two layout
# refusals, §5.9 #7, #15).
counters_zero? = fn report ->
  report |> Map.drop([:malformed, :layout_hash]) |> Map.values() |> Enum.all?(&(&1 == 0))
end

{:ok, checker} = Agent.start_link(fn -> {0, []} end)

check = fn label, pass? ->
  Agent.get_and_update(checker, fn {n, fails} ->
    if pass? do
      IO.puts("ok #{label}")
      {n, {n + 1, fails}}
    else
      IO.puts("FAIL #{label}")
      {n, {n + 1, fails ++ [label]}}
    end
  end)
end

named_refusal = fn label, data, want, want_layout_hash ->
  case load.(data) do
    {:error, ^want, report} ->
      check.(
        label,
        not report.malformed and counters_zero?.(report) and
          report.layout_hash == want_layout_hash
      )

    other ->
      check.(label <> " — wanted {:error, #{inspect(want)}, _}, got #{inspect(other)}", false)
  end
end

malformed_read = fn label, data ->
  case load.(data) do
    {:error, :malformed, report} ->
      check.(label, report.malformed and counters_zero?.(report) and report.layout_hash == 0)

    other ->
      check.(
        label <>
          " — wanted {:error, :malformed, _} with malformed set, got #{inspect(other)}",
        false
      )
  end
end

# ---- the positive control: a lawful read lands, malformed clear ----

own_file = header.(hash, layout) <> layout <> record.(hash)

check.(
  "lawful own-file read lands {:ok, values, report}",
  match?({:ok, values, _} when is_list(values), load.(own_file))
)

case load.(own_file) do
  {:ok, _values, report} ->
    check.("lawful own-file read: malformed clear", not report.malformed)

  other ->
    check.(
      "lawful own-file read: malformed clear — got #{inspect(other)}",
      false
    )
end

# ---- REFUSAL BY NAME: refused true, reason the name, malformed FALSE ----
# (docs/FIXED-FORM-ALGORITHM.md:906, the table's first row)

# form byte 1 — a committed form-1 file, whatever its length
named_refusal.(
  "form byte 1 refuses previous_form, malformed clear, no counter",
  <<1, 0, 0>>,
  :previous_form,
  0
)

# form byte 2 — a committed form-2 batch
named_refusal.(
  "form byte 2 refuses message_form_as_file, malformed clear, no counter",
  <<2, 0, 0>>,
  :message_form_as_file,
  0
)

# a form byte no form defines
named_refusal.(
  "form byte 9 refuses newer_form, malformed clear, no counter",
  <<9, 0, 0>>,
  :newer_form,
  0
)

# a hash no lineage entry holds — layout_newer reports THE FILE'S HASH and
# nothing else (§5.9 #7)
stranger = Bitwise.bxor(hash, 0xDEADBEEF)

check.(
  "the stranger hash is in no lineage entry (vector soundness)",
  not Enum.any?(Tuple.to_list(known), &(&1.hash == stranger))
)

named_refusal.(
  "a hash outside the lineage refuses layout_newer, layout_hash the file's hash, no counter",
  header.(stranger, layout) <> layout <> record.(stranger),
  :layout_newer,
  stranger
)

# a KNOWN hash whose layout bytes differ — layout_malformed, the name, and
# layout_hash stays 0 (zero on every path but the two layout refusals, §5.9 #15)
corrupt =
  binary_part(layout, 0, byte_size(layout) - 1) <>
    <<Bitwise.bxor(:binary.at(layout, byte_size(layout) - 1), 1)>>

named_refusal.(
  "a known hash over different layout bytes refuses layout_malformed, layout_hash clear",
  header.(hash, corrupt) <> corrupt <> record.(hash),
  :layout_malformed,
  0
)

# a record whose hash names no layout this reader holds
named_refusal.(
  "a record hash that is not the file's refuses no_layout, malformed clear, no counter",
  header.(hash, layout) <> layout <> record.(Bitwise.bxor(hash, 1)),
  :no_layout,
  0
)

# ---- MALFORMED: the residue, malformed TRUE, no reason named ----
# (docs/FIXED-FORM-ALGORITHM.md:906, the table's second row)

# no first byte — there is no form byte to name
malformed_read.("an empty file is malformed, counters zero", <<>>)

# under 20 bytes, the form byte having already answered
malformed_read.("a 5-byte form-3 file is malformed, counters zero", <<3, 0, 0, 0, 0>>)

# any reserved byte 1..7 nonzero — the reserved clause stands before the length
# clause, so eight bytes are enough to name it
malformed_read.("a nonzero reserved byte is malformed, counters zero", <<3, 0, 0, 0, 0, 0, 0, 1>>)

# a ragged tail: rest mod record_bytes != 0
malformed_read.("a ragged tail is malformed, counters zero", own_file <> <<1, 2, 3>>)

{count, fails} = Agent.get(checker, & &1)
Agent.stop(checker)

if fails == [] do
  IO.puts("#{count} assertions, all green — REFUSE is total on the #{mod} read")
  System.halt(0)
else
  IO.puts(:stderr, "#{count} assertions, #{length(fails)} failed: #{Enum.join(fails, "; ")}")
  System.halt(1)
end
